packages/web/src/components/createCell.tsx

```diff
@@ -131,14 +133,18 @@ export function createCell<CellProps = any>({
             } else {
               throw error
             }
-          } else if (loading) {
-            return <Loading {...queryRest} {...props} />
           } else if (data) {
-            if (typeof Empty !== 'undefined' && isEmpty(data)) {
+            if (typeof Updating !== 'undefined' && loading) {
+              return (
+                <Updating {...afterQuery(data)} {...queryRest} {...props} />
```

Comment:
My main concern with this is developer implementation and potential re-renders.  

From my understanding, The reason we want to implement this is when there is already data in the cache and apollo re-fetches we see the loading component instead of the cached data.

This is opt-in and will almost be a direct duplication to the Success Component with maybe some extra props. Does it make more sense to pass updating prop to success? Since `Updating` is its own component it will not just re-render the `Posts` component it will unmount it first and then render it. This could cause issues if there are subsequent requests / Cells inside of the `Posts` component

```javascript
const Updating = ({ updating, posts}) => {
     return <Posts  updating={updating} posts={posts}  />
}

or 

const Updating = ({ updating, posts }) => {
     return (
     <>
        {updating && <Spinner />}
         <Posts posts={posts} />
    </>

}
```

```javascript
const Success = ({ posts, updating }) => {
    return <Posts posts={posts} />
}
```

Reply:
Doesn't React only do a redraw if it diffs and sees that *something* will be different based on the data inputs? If the input is the same then it shouldn't try to update the DOM? Or maybe just by virtue of it being a different component it will ALWAYS re-render?

I just deployed my site with a patch that adds this behavior and I'm not seeing any kind of visual indication that a re-render is happening...are you more worried about CPU/memory in general?

Reply:
I think @KrisCoulson is right, and what you said here @cannikin is spot on:

> Or maybe just by virtue of it being a different component it will ALWAYS re-render?

This is the corresponding section of the docs: https://reactjs.org/docs/reconciliation.html#elements-of-different-types.

Reply:
React's flamegraph definitely says that Success remounts after the Cell goes from Success -> Updating. But it looks like nothing happens. I even slowed it down: https://s.tape.sh/r7eFQbeB. Can't say that's true of every case; I'll keep stress testing it.

Reply:
One way we could handle this (but it comes with some drawbacks (i.e. your Updating can't render something totally different from Empty/Success—but would you want to render something totally different?)):

```js
export const Success = ({ posts, refetch, Updating }) => {
  return (
    <>
      <Posts posts={posts} />
      <button onClick={() => refetch()}>refetch</button>
      <Updating />
    </>
  )
}

export const Updating = () => <Loading />
```

What's going on in the example above is that Updating is passed as a prop to Success. The difference is, Updating automatically renders nothing (i.e. returns `null`) when `loading = false`. Here's what I mean (this is what the Cell code would look like):

```js
<Success
  {...afterQuery(data)}
  {...queryRest}
  {...props}
  Updating={
    loading
      ? (props: SuperUncoolTypez) => <Updating {...props} />
      : () => null
    }
/>
```

The great thing about this is that Empty is handled: it gets Updating too.

We could still provide `loading` to Empty/Success, just so you can always do whatever you want.

Reply:
Nothing gets unmounted this way:


https://user-images.githubusercontent.com/32992335/119763620-be153500-be64-11eb-87eb-4a2225018758.mp4